cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/RemotePreferencesFetcherInternal.java

```diff
@@ -0,0 +1,115 @@
+ */
+public class RemotePreferencesFetcherInternal implements AbstractPreferencesFetcher {
+  private static final Gson GSON = new Gson();
+
+  private final RemoteClient remoteClient;
+
+  @Inject
+  public RemotePreferencesFetcherInternal(final DiscoveryServiceClient discoveryClient) {
+    this.remoteClient = new RemoteClient(
+      discoveryClient, Constants.Service.APP_FABRIC_HTTP,
+      new DefaultHttpRequestConfig(false), Constants.Gateway.INTERNAL_API_VERSION_3);
+  }
+
+  /**
+   * Get preferences for the given identify
+   */
+  public PreferencesDetail get(EntityId entityId, boolean resolved) throws IOException, NotFoundException {
+    HttpResponse httpResponse;
+    String url = getPreferencesURI(entityId, resolved);
+    HttpRequest.Builder requestBuilder = remoteClient.requestBuilder(HttpMethod.GET, url);
+    try {
+      httpResponse = execute(requestBuilder.build());
+    } catch (NotFoundException e) {
+      throw new NotFoundException(entityId);
+    }
+    return GSON.fromJson(httpResponse.getResponseBodyAsString(), new TypeToken<PreferencesDetail>() {
+    }.getType());
+  }
+
+  /**
+   * Construct URI to fetch preferences depending on the type of supplied entity
+   */
+  private String getPreferencesURI(EntityId entityId, boolean resolved) {
+    String uri;
+    switch (entityId.getEntityType()) {
+      case INSTANCE:
+        uri = String.format("preferences");
+        break;
+      case NAMESPACE:
+        NamespaceId namespaceId = (NamespaceId) entityId;
+        uri = String.format("namespaces/%s/preferences", namespaceId.getNamespace());
+        break;
+      case APPLICATION:
+        ApplicationId appId = (ApplicationId) entityId;
+        uri = String.format("namespaces/%s/apps/%s/preferences",
+                            appId.getNamespace(), appId.getApplication());
+        break;
+      case PROGRAM:
+        ProgramId programId = (ProgramId) entityId;
+        uri = String.format("namespaces/%s/apps/%s/%s/%s/preferences",
+                            programId.getNamespace(), programId.getApplication(), programId.getType().getCategoryName(),
+                            programId.getProgram());
+        break;
+      default:
+        throw new UnsupportedOperationException(
+          String.format("Preferences cannot be used on this entity type: %s", entityId.getEntityType()));
+    }
+    if (resolved) {
+      uri += "?resolved=true";
+    }
+    return uri;
+  }
+
+  private HttpResponse execute(HttpRequest request) throws IOException, NotFoundException {
+    HttpResponse httpResponse = remoteClient.execute(request);
+    if (httpResponse.getResponseCode() == HttpURLConnection.HTTP_NOT_FOUND) {
+      throw new NotFoundException("Not found");
+    }
+    if (httpResponse.getResponseCode() != HttpURLConnection.HTTP_OK) {
+      throw new IOException(String.format("Request failed %s", httpResponse.getResponseBodyAsString()));
```

Comment:
there is code duplication in these classes, it would be better to consolidate the common code, either in a common abstract class or by enhancing RemoteClient in some way.

Reply:
Agreed. Added a TODO to refactor in a follow-up PR